wsmaster/che-core-api-agent/src/main/java/org/eclipse/che/api/agent/server/impl/LocalAgentRegistryImpl.java

```diff
@@ -0,0 +1,121 @@
+package org.eclipse.che.api.agent.server.impl;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import org.eclipse.che.api.agent.server.AgentRegistry;
+import org.eclipse.che.api.agent.server.exception.AgentException;
+import org.eclipse.che.api.agent.server.exception.AgentNotFoundException;
+import org.eclipse.che.api.agent.shared.dto.AgentDto;
+import org.eclipse.che.api.agent.shared.model.Agent;
+import org.eclipse.che.api.agent.shared.model.AgentKey;
+import org.eclipse.che.commons.lang.IoUtil;
+import org.eclipse.che.dto.server.DtoFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static java.lang.String.format;
+import static java.util.Collections.singleton;
+import static java.util.Collections.unmodifiableSet;
+import static org.eclipse.che.commons.lang.IoUtil.downloadFile;
+import static org.eclipse.che.commons.lang.IoUtil.readAndCloseQuietly;
+
+/**
+ * Local implementation of the {@link AgentRegistry}.
+ * The name of the agent might represent a url.
+ *
+ * @author Anatoliy Bazko
+ */
+@Singleton
+public class LocalAgentRegistryImpl implements AgentRegistry {
+
+    protected static final Logger  LOG    = LoggerFactory.getLogger(LocalAgentRegistryImpl.class);
+    private static final   Pattern AGENTS = Pattern.compile(".*[//]?agents/[^//]+[.]json");
+
+    private final Map<String, Agent>     agents;
+
+    @Inject
+    public LocalAgentRegistryImpl() throws IOException {
+        this.agents = new HashMap<>();
+
+        IoUtil.getResources(this.getClass(), AGENTS, inputStream -> {
+            final Agent agent;
+            try {
+                agent = DtoFactory.getInstance().createDtoFromJson(inputStream, AgentDto.class);
+                agents.put(agent.getName(), agent);
+            } catch (IOException e) {
+                LOG.error("Can't create agent.", e);
+            }
+        });
+    }
+
+    @Override
+    public Agent createAgent(String name, String version) throws AgentException {
+        return doCreateAgent(name);
+    }
+
+    @Override
+    public Agent createAgent(AgentKey agentKey) throws AgentException {
+        return doCreateAgent(agentKey.getName());
+    }
+
+    @Override
+    public Agent createAgent(String name) throws AgentException {
+        return doCreateAgent(name);
+    }
+
+    @Override
+    public Collection<String> getVersions(String name) throws AgentException {
+        Agent agent = doCreateAgent(name);
+        return singleton(agent.getVersion());
+    }
+
+    @Override
+    public Set<String> getAgents() throws AgentException {
+        return unmodifiableSet(agents.keySet());
```

Comment:
it looks like `agents` map is not modifiable at all, so what do you think about using instance of `ImmutableMap` instead, then you don't need to wrap `keySet()` with a new unmodifiableSet instance each time `getAgents()` is called